src/arch/riscv64/interrupts.rs

```diff
@@ -2,8 +2,122 @@ use crate::arch::riscv64::registers;
 use crate::arch::ArchitectureInterrupts;
 use crate::drivers::plic::plic_handler;
 use crate::kprintln;
+
 use core::arch::asm;
 
+use riscv;
+use sbi;
+
+#[repr(u64)]
+#[derive(Debug, Copy, Clone)]
+enum InterruptType {
+    Reserved,
+    SupervisorSoftware,
+    SupervisorTimer,
+    SupervisorExternal,
+    Platform(u64),
+}
+
+impl InterruptType {
+    fn is_asynchronous(&self) -> bool {
+        matches!(self, Self::SupervisorTimer)
+    }
+}
+
+impl From<u64> for InterruptType {
+    fn from(code: u64) -> Self {
+        match code {
+            0 | 2..=4 | 6..=8 | 10..=15 => Self::Reserved,
+            1 => Self::SupervisorSoftware,
+            5 => Self::SupervisorTimer,
+            9 => Self::SupervisorExternal,
+            _ => Self::Platform(code),
+        }
+    }
+}
+
+impl From<InterruptType> for u64 {
+    fn from(itype: InterruptType) -> Self {
+        match itype {
```

Comment:
So we can't do `itype as u64` here even tho it's `repr(u64)`?

Reply:
I should remove the `repr(u64)`, this is legacy of the `enum` whit assign discriminant.

The `itype as u64` was no working with the nightly feature required for discriminant with `enum` containing sum types. There was some ugly unsafe code so I remove it.

Reply:
Fixed by [e60eb21](https://github.com/ks0n/goOSe/pull/95/commits/e60eb21f12707f423ea16cbc00cadec80b88fb5b)